clang/test/SemaSYCL/intel-fpga-local-ast.cpp

```diff
@@ -0,0 +1,537 @@
+
+  // Checking of duplicate argument values.
+  //CHECK: VarDecl{{.*}}var_forcep2d
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelForcePow2DepthAttr
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}1
+  //CHECK-NEXT: IntegerLiteral{{.*}}1{{$}}
+  [[intel::force_pow2_depth(1)]]
+  [[intel::force_pow2_depth(1)]] int var_forcep2d; // OK
+
+  // Test for Intel FPGA bankwidth memory attribute duplication.
+  // No diagnostic is emitted because the arguments match.
+  // Duplicate attribute is silently ignored.
+  //CHECK: VarDecl{{.*}}var_bankwidth 'int'
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit Default
+  //CHECK: SYCLIntelBankWidthAttr
+  //CHECK-NEXT: ConstantExpr{{.*}}'int'
+  //CHECK-NEXT: value: Int 16
+  //CHECK-NEXT: IntegerLiteral{{.*}}'int' 16
+  [[intel::bankwidth(16)]]
+  [[intel::bankwidth(16)]] int var_bankwidth; // OK
+
+  // Test for Intel FPGA numbanks memory attribute duplication.
+  // No diagnostic is emitted because the arguments match.
+  // Duplicate attribute is silently ignored.
+  //CHECK: VarDecl{{.*}}var_numbanks 'int'
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit Default
+  //CHECK: SYCLIntelNumBanksAttr
+  //CHECK-NEXT: ConstantExpr{{.*}}'int'
+  //CHECK-NEXT: value: Int 8
+  //CHECK-NEXT: IntegerLiteral{{.*}}'int' 8
+  [[intel::numbanks(8)]]
+  [[intel::numbanks(8)]] int var_numbanks; // OK
+
+  // Checking of different argument values.
+  //CHECK: VarDecl{{.*}}bw_bw 'unsigned int[64]'
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit Default
+  //CHECK: SYCLIntelBankWidthAttr
+  //CHECK-NEXT: ConstantExpr{{.*}}'int'
+  //CHECK-NEXT: value: Int 8
+  //CHECK-NEXT: IntegerLiteral{{.*}}'int' 8
+  //expected-warning@+2{{attribute 'bankwidth' is already applied}}
+  [[intel::bankwidth(8)]] // expected-note {{previous attribute is here}}
+  [[intel::bankwidth(16)]] unsigned int bw_bw[64];
+
+  //CHECK: VarDecl{{.*}}nb_nb 'unsigned int[64]'
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit Default
+  //CHECK: SYCLIntelNumBanksAttr
+  //CHECK-NEXT: ConstantExpr{{.*}}'int'
+  //CHECK-NEXT: value: Int 8
+  //CHECK-NEXT: IntegerLiteral{{.*}}'int' 8
+  //expected-warning@+2{{attribute 'numbanks' is already applied}}
+  [[intel::numbanks(8)]] // expected-note {{previous attribute is here}}
+  [[intel::numbanks(16)]] unsigned int nb_nb[64];
+
+  //Last one is applied and others ignored.
+  //CHECK: VarDecl{{.*}}mrg_mrg
+  //CHECK: SYCLIntelMergeAttr{{.*}}"mrg4" "depth"{{$}}
+  //CHECK: SYCLIntelMergeAttr{{.*}}"mrg5" "width"{{$}}
+  //expected-warning@+2{{attribute 'merge' is already applied}}
+  [[intel::merge("mrg4", "depth")]]
+  [[intel::merge("mrg5", "width")]] unsigned int mrg_mrg[4];
+
+  //CHECK: VarDecl{{.*}}bb_bb
+  //CHECK: SYCLIntelBankBitsAttr
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}42
+  //CHECK-NEXT: IntegerLiteral{{.*}}42{{$}}
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}43
+  //CHECK-NEXT: IntegerLiteral{{.*}}43{{$}}
+  //CHECK: SYCLIntelBankBitsAttr
```

Comment:
Ok. So in all of these tests I see duplicate attributes being applied after the warning. Is that correct behavior? Shouldn't we be ignoring duplicate attributes?

Reply:
Thanks @elizabethandrews for reviews. This is a problem with existing implementation of SYCLIntelBankBitsAttr. Since i am not making any functionality changes in this PR. i have created Github issue https://github.com/intel/llvm/issues/7817 to address this problem in a separate patch. Are you OK with it?

Reply:
Done